solidity/test/token_stake/TestTopUps.js

```diff
@@ -0,0 +1,676 @@
+    })
+  })
+
+  describe("escrow redelegation top-ups", async () => {
+    beforeEach(async () => {
+      await time.increase(initializationPeriod.addn(1))
+      await tokenStaking.undelegate(operatorTwo, {from: grantee})
+      await tokenStaking.undelegate(operatorThree, {from: managedGrantee})
+      await time.increase(undelegationPeriod.addn(1))
+      await tokenStaking.recoverStake(operatorTwo)
+      await tokenStaking.recoverStake(operatorThree)
+    })
+
+    it("can be done for a grant", async () => {
+      await delegateStakeFromGrant(
+        tokenGrant,
+        tokenStaking.address,
+        grantee,
+        operatorFour,
+        beneficiary,
+        authorizer,
+        delegatedAmount,
+        grantId
+      )
+      await time.increase(initializationPeriod.addn(1))
+
+      let delegationInfo = await tokenStaking.getDelegationInfo(operatorFour)
+      expect(delegationInfo.amount).to.eq.BN(delegatedAmount)
+
+      const data = Buffer.concat([
+        Buffer.from(beneficiary.substr(2), 'hex'),
+        Buffer.from(operatorFour.substr(2), 'hex'),
+        Buffer.from(authorizer.substr(2), 'hex')
+      ])
+      await tokenStakingEscrow.redelegate(
+        operatorTwo, delegatedAmount, data, {from: grantee}
+      )
+
+      await time.increase(initializationPeriod.addn(1))
+      delegationInfo = await tokenStaking.getDelegationInfo(operatorFour)
+      expect(delegationInfo.amount).to.eq.BN(delegatedAmount)
+  
+      await tokenStaking.commitTopUp(operatorFour, {from: grantee})
+        
+      delegationInfo = await tokenStaking.getDelegationInfo(operatorFour)
+      expect(delegationInfo.amount).to.eq.BN(delegatedAmount.muln(2))
+    })
+
+    it("can be done for a managed grant", async () => {
+      await delegateStakeFromManagedGrant(
+        managedGrant,
+        tokenStaking.address,
+        managedGrantee,
+        operatorFour,
+        beneficiary,
+        authorizer,
+        delegatedAmount
+      ) 
+      await time.increase(initializationPeriod.addn(1))
+
+      let delegationInfo = await tokenStaking.getDelegationInfo(operatorFour)
+      expect(delegationInfo.amount).to.eq.BN(delegatedAmount)
+
+      const data = Buffer.concat([
+        Buffer.from(beneficiary.substr(2), 'hex'),
+        Buffer.from(operatorFour.substr(2), 'hex'),
+        Buffer.from(authorizer.substr(2), 'hex')
+      ])
+      await tokenStakingEscrow.redelegate(
+        operatorThree, delegatedAmount, data, {from: managedGrantee}
+      )
+
+      await time.increase(initializationPeriod.addn(1))
+      delegationInfo = await tokenStaking.getDelegationInfo(operatorFour)
+      expect(delegationInfo.amount).to.eq.BN(delegatedAmount)
+  
+      await tokenStaking.commitTopUp(operatorFour, {from: managedGrantee})
+        
+      delegationInfo = await tokenStaking.getDelegationInfo(operatorFour)
+      expect(delegationInfo.amount).to.eq.BN(delegatedAmount.muln(2))
+    })
+  })
```

Comment:
Can we add a test case to cover `No top up to commit` check?

Reply:
Done in https://github.com/keep-network/keep-core/pull/1893/commits/870a6caf5187c43e5fd2d48f9aa6540d31e4f19c